core/src/main/scala/org/apache/spark/sql/delta/commands/CloneTableBase.scala

```diff
@@ -0,0 +1,356 @@
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.execution.command.LeafRunnableCommand
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.{Clock, SerializableConfiguration}
+// scalastyle:on import.ordering.noEmptyLine
+
+/**
+ * An interface of the source table to be cloned from.
+ */
+trait CloneSource extends Closeable {
+  /** The format of the source table */
+  def format: String
+
+  /** The source table's protocol */
+  def protocol: Protocol
+
+  /** A system clock */
+  def clock: Clock
+
+  /** The source table name */
+  def name: String
+
+  /** The path of the source table */
+  def dataPath: Path
+
+  /** The source table schema */
+  def schema: StructType
+
+  /** The catalog table of the source table, if exists */
+  def catalogTable: Option[CatalogTable]
+
+  /** The time travel spec of the source table, if exists */
+  def timeTravelOpt: Option[DeltaTimeTravelSpec]
+
+  /** A snapshot of the source table, if exists */
+  def snapshot: Option[Snapshot]
+
+  /** The metadata of the source table */
+  def metadata: Metadata
+
+  /** All of the files present in the source table */
+  def allFiles: Dataset[AddFile]
+
+  /** Total size of data files in bytes */
+  def sizeInBytes: Long
+
+  /** Total number of data files */
+  def numOfFiles: Long
+
+  /** Describe this clone source */
+  def description: String
+}
+
+// Clone source table formats
+object CloneSourceFormat {
+  val DELTA = "Delta"
+  val PARQUET = "Parquet"
+  val UNKNOWN = "Unknown"
+}
+
+trait CloneTableBaseUtils extends DeltaLogging
+{
+
+  import CloneTableCommand._
+
+  /** Make a map of operation metrics for the executed command for DeltaLog commits */
+  protected def getOperationMetricsForDeltaLog(
+      opMetrics: SnapshotOverwriteOperationMetrics): Map[String, Long] = {
+    Map(
+      SOURCE_TABLE_SIZE -> opMetrics.sourceSnapshotSizeInBytes,
+      SOURCE_NUM_OF_FILES -> opMetrics.sourceSnapshotFileCount,
+      NUM_REMOVED_FILES -> opMetrics.destSnapshotRemovedFileCount,
+      NUM_COPIED_FILES -> 0L,
+      REMOVED_FILES_SIZE -> opMetrics.destSnapshotRemovedFilesSizeInBytes,
+      COPIED_FILES_SIZE -> 0L)
+  }
+
+  /**
+   * Make a map of operation metrics for the executed command for recording events.
+   * Any command can extend to overwrite or add new metrics
+   */
+  protected def getOperationMetricsForEventRecord(
+      opMetrics: SnapshotOverwriteOperationMetrics): Map[String, Long] =
+    getOperationMetricsForDeltaLog(opMetrics)
+
+  /** Make a output Seq[Row] of metrics for the executed command */
+  protected def getOutputSeq(operationMetrics: Map[String, Long]): Seq[Row]
+
+
+  protected val fnfExceptionMessage: String = "Tried to clone a version of the table where " +
```

Comment:
where is this used? Do we check for the existence of files in source table before the clone?

Reply:
yeah we do, but this error message is not useful any more.